src/scipp/core/bin_remapping.py

```diff
@@ -1,114 +1,139 @@
         # buffer copy. Instead index just begin/end and reuse content buffer.
         comps = da.bins.constituents
-        select = ~mask
-        data = _cpp._bins_no_validate(
-            data=comps['data'],
-            dim=comps['dim'],
-            begin=comps['begin'][select],
-            end=comps['end'][select],
-        )
+        # If the mask is 1-D we can drop entire "rows" or "columns". This can
+        # drastically reduce the number of bins to handle in some cases for better
+        # performance. For 2-D or higher masks we fall back to making bins "empty" by
+        # setting end=begin.
+        if mask.ndim == 1:
+            select = ~mask
+            comps['begin'] = comps['begin'][select]
+            comps['end'] = comps['end'][select]
+        else:
+            comps['end'] = where(mask, comps['begin'], comps['end'])
+        return _cpp._bins_no_validate(**comps)
     else:
-        data = da.data
-    return DataArray(data,
-                     coords=reduced_coords(da, dim),
-                     masks=reduced_masks(da, dim),
-                     attrs=reduced_attrs(da, dim))
-
-
-def _concat_bins_variable(var: Variable, dim: str) -> Variable:
-    # We want to write data from all bins along dim to a contiguous chunk in the
-    # content buffer. This will then allow us to create new, larger bins covering the
-    # respective input bins. We use `cumsum` after moving `dim` to the innermost dim.
-    # This will allow us to setup offsets for the new contiguous layout.
-    sizes = var.bins.size()
-    out_dims = [d for d in var.dims if d != dim]
-    out_end = cumsum(sizes.transpose(out_dims + [dim])).transpose(var.dims)
-    out_begin = out_end - sizes
-    out = _cpp._bins_no_validate(
-        data=copy_for_overwrite(var.bins.constituents['data']),
+        return da.data
+
+
+def _with_bin_sizes(var: Variable, sizes: Variable) -> Variable:
+    end = cumsum(sizes)
+    begin = end - sizes
+    data = var if var.bins is None else var.bins.constituents['data']
+    dim = var.dim if var.bins is None else var.bins.constituents['dim']
+    return _cpp._bins_no_validate(data=data, dim=dim, begin=begin, end=end)
+
+
+def _concat_bins(var: Variable, dim: List[str]) -> Variable:
+    # To concat bins, two things need to happen:
+    # 1. Data needs to be written to a contiguous chunk.
+    # 2. New bin begin/end indices need to be setup.
+    # If the dims to concatenate are the *inner* dims a call to `copy()` performs 1.
+    # Otherwise, we first transpose and then `copy()`.
+    # For step 2. we simply sum the (transposed) input bin sizes over the concat dims,
+    # which `_with_bin_sizes` can use to compute new begin/end indices.
+    changed_dims = list(concrete_dims(var, dim))
+    unchanged_dims = [d for d in var.dims if d not in changed_dims]
+    # TODO It would be possible to support a copy=False parameter, to skip the copy if
+    # the copy would not result in any moving or reordering.
```

Comment:
Nothing really, feel free to pick it up if you feel like it, as a follow-up.

Reply:
I am rather asking to avoid putting in more TODO comments. Converting this to an issue so we don't forget.

Reply:
I don't think it would be bad to forget: Everything works, so unless someone has performance issues (and investigates the code...).

Generally, I do not think GitHub issues should be used to write done things that "could be done in principle". Unless we have a concrete plan to actually do it, it just makes the issue backlog less useful.